sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java

```diff
@@ -0,0 +1,311 @@
+
+                    return exp.invoke("toString", TypeDef.of(String.class)).returning();
+                })
+            );
+        classDefBuilder.addMethod(method);
+    }
+
+    /*
+    Creates an equals method with signature:
+        public static boolean BeanNameObject.equals(BeanName object1, Object object2)
+     */
+    private static void createEqualsMethod(ClassDef.ClassDefBuilder classDefBuilder, String objectName, List<PropertyElement> properties) {
+        MethodDef method = MethodDef.builder("equals")
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .returns(boolean.class)
+            .addParameter("instance", ClassTypeDef.of(objectName))
+            .addParameter("o", TypeDef.of(Object.class))
+            .build((self, parameterDef) -> {
+                VariableDef instance = parameterDef.get(0).asVariable();
+                VariableDef o = parameterDef.get(1).asVariable();
+                return StatementDef.multi(
+                    instance.asCondition(" == ", o)
+                        .asConditionIf(ExpressionDef.trueValue().returning()),
+                    o.isNull().asCondition(" || ",
+                        instance.invoke("getClass", ClassTypeDef.of("Class"))
+                            .asCondition(" != ", o.invoke("getClass", ClassTypeDef.of("Class"))))
+                        .asConditionIf(ExpressionDef.falseValue().returning()),
+                    o.cast(ClassTypeDef.of(objectName)).newLocal("other", variableDef -> {
+                        ExpressionDef exp = null;
+                        TypeDef propertyTypeDef;
+                        ExpressionDef.CallInstanceMethod firstProperty;
+                        ExpressionDef.CallInstanceMethod secondProperty;
+                        for (PropertyElement beanProperty : properties) {
+                            propertyTypeDef = TypeDef.of(beanProperty.getType());
+                            if ( !beanProperty.hasAnnotation(EqualsAndHashCode.Exclude.class) && beanProperty.getReadMethod().isPresent()) {
+                                firstProperty = instance.invoke(beanProperty.getReadMethod().get(), List.of());
+                                secondProperty = variableDef.invoke(beanProperty.getReadMethod().get(), List.of());
+                            } else {
+                                continue;
+                            }
+                            ExpressionDef newEqualsExpression = propertyTypeDef.isPrimitive() ?
+                                firstProperty.asCondition(" == ", secondProperty)
+                                : firstProperty.asCondition(" == ", secondProperty)
+                                .asConditionOr(firstProperty.isNonNull().asConditionAnd(
+                                    (propertyTypeDef.isArray()) ?
+                                        ClassTypeDef.of(Arrays.class).invokeStatic("equals", TypeDef.BOOLEAN, Arrays.asList(firstProperty, secondProperty))
```

Comment:
Why is it a list of values and not just `"equals", TypeDef.BOOLEAN, firstProperty, secondProperty`?

Reply:
Because there is support for both and (ExpressionDef... parameters) gets cast into a List anyway. But I can change